doc/man/gnunet-qr.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd September 13, 2014
.Dt GNUNET-QR 1
.Os
.Sh NAME
.Nm gnunet-qr
.Nd scan a QR code using a video device and import
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl d Ar DEVICE | Fl -device= Ns Ar DEVICE
.Op Fl h | -help
.Op Fl s | -silent
.Op Fl v | -verbose
.Sh DESCRIPTION
.Nm
is a command line tool to scan a QR code using a video device and import.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file
.Ar FILENAME .
.It Fl d Ar DEVICE | Fl -device= Ns Ar DEVICE
Use device
.Ar DEVICE .
.It Fl h | -help
Print short help on options.
.It Fl s | -silent
Do not show preview windows.
.It Fl v | -verbose
Be verbose.
.El
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100
  utils: gnunet-qr.py now installed under pkgshare instead of bin

Thu Feb 28 19:07:59 CET 2019
  CADET: Fix in order, unrealiable message ordering

Thu 28 Feb 2019 10:54:58 AM CET
  Releasing GNUnet 0.11.0. -CG

Wed Jun  6 08:28:27 CEST 2018
  Releasing GNUnet 0.11.0pre66. Let's try to keep a better
  ChangeLog again henceforth. -CG

Sat Jan  7 16:41:34 CET 2017
  Converting NAT library to new NAT service (and
  splitting of nat-auto service for auto-configuration). -CG


src/util/gnunet-qr.py
import sys
import getopt
import subprocess
from sys import argv
try:
    import zbar
except ImportError as e:
    print('Cannot run gnunet-qr, please install the zbar module.')
    print('For Debian, you can obtain it as "python-zbar".')
    print('Upstream: http://zbar.sourceforge.net/')
    sys.exit(1)


def help():
    print('gnunet-qr\n\
Scan a QR code using a video device and import\n\
Arguments mandatory for long options are also mandatory for short options.\n\
  -c, --config FILENAME      use configuration file FILENAME\n\
  -d, --device DEVICE        use device DEVICE\n\
  -s, --silent               do not show preview windows\n\
  -h, --help                 print this help\n\
  -v, --verbose              be verbose\n\
Report bugs to gnunet-developers@gnu.org.\n\
GNUnet home page: https://gnunet.org/\n\
General help using GNU software: https://www.gnu.org/gethelp/')


if __name__ == '__main__':
    configuration = ''
    device = '/dev/video0'
    url = ''
    verbose = False
    silent = False
    # Parse arguments
    try:
        opts, args = getopt.gnu_getopt(sys.argv[1:], "c:hd:sv", ["config", "help", "device", "silent", "verbose"])
    except getopt.GetoptError as e:
        help()
        print(str(e))
        exit(1)
    for o, a in opts:
        if o in ("-h", "--help"):
            help()
            sys.exit(0)
        elif o in ("-c", "--config"):
            configuration = a
        elif o in ("-d", "--device"):
            device = a
        elif o in ("-s", "--silent"):
            silent = True
        elif o in ("-v", "--verbose"):
            verbose = True
    if (True == verbose):
        print('Initializing')
    # create a Processor
    proc = zbar.Processor()

    # configure the Processor
    proc.parse_config('enable')

    # initialize the Processor
    try:
        if (True == verbose):
            print('Opening video device ' + device)
        proc.init(device)
    except Exception as e:
        print('Failed to open device ' + device)
        exit(1)

    # enable the preview window
    # if (True == silent):
    #       proc.visible = True
    # else:
    #               proc.visible = False

    proc.visible = True
    # read at least one barcode (or until window closed)
    try:
        if (True == verbose):
            print('Capturing')
        proc.process_one()
    except Exception as e:
        # Window was closed without finding code
        exit(1)

    # hide the preview window
    proc.visible = False

    # extract results
    for symbol in proc.results:
        # do something useful with results
        if (True == verbose):
            print('Found ', symbol.type, ' symbol ', '"%s"' % symbol.data)
        args = list()
        args.append("gnunet-uri")
        if (configuration != ''):
            args.append(str("-c " + str(configuration)))
        args.append(str(symbol.data))
        cmd = ''
        for a in args:
            cmd += " " + str(a)
        if (verbose):
            print('Running `' + cmd +'`')
        res = subprocess.call(args)
        if (0 != res):
            print('Failed to add URI ' + str(symbol.data))
        else:
            print('Added URI ' + str(symbol.data))
        exit(res)
    exit(1)
